### PR TITLE
fix(framework): Ensure steps after skipped steps are executed

### PR DESCRIPTION
**Pull request by rifont**
### What changed? Why was the change needed?

* Ensure steps after skipped steps are executed
  * added tests to verify the bugfix
* Add tests to ensure steps resolve functions after the executed step are not called
* Rename the early exit promise function and add relevant comments
* Remove redundant event cloning

### Screenshots

_**Workflow under test**`_
```typescript
export const doubleSkipWorkflow = workflow(
  "double-skip-workflow",
  async ({ step, payload }) => {
    console.log("payload ==>", payload);
    await step.email(
      "send-email",
      () => ({
        subject: "Your verification code",
        body: `Hello! This is your verification code: ${payload.code}`,
      }),
      {
        skip: () => payload.type !== "email",
      }
    );

    await step.sms(
      "send-sms",
      () => {
        return {
          body: `Your verification code is: ${payload.code}`,
        }
      },
      {
        skip: () => payload.type !== "sms",
      }
    );
  },
  {
    payloadSchema: z.object({
      code: z.string(),
      type: z.enum(["email", "sms"]).default("email"),
    }),
  }
);

await doubleSkipWorkflow.trigger({ payload: { code: '123456', type: 'sms' } });
```

_BEFORE - steps after a skipped step were also skipped when `skip` condition was not met_
image

_AFTER - steps after a skipped step are executed when `skip` condition is not met_
image

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/framework/src/client.ts [hunk 1]
```diff
@@ -261,14 +260,23 @@ export class Client {
   ): ActionStep<T_Outputs, T_Result> {
     return async (stepId, stepResolve, options) => {
       const step = this.getStep(event.workflowId, stepId);
-      const eventClone = cloneData(event);
```
Comment (rifont) [inline packages/framework/src/client.ts:264]: Event cloning is no longer needed as the problematic `ajv` validator which mutated the data is now cloning internally.

File: packages/framework/src/client.ts [hunk 2]
```diff
@@ -347,13 +355,25 @@ export class Client {
...
+     * `workflow.execute` method. By resolving the `concludeExecutionPromise` when setting the result,
+     * we can ensure that the `workflow.execute` method is not evaluated after the `stepId` is reached.
+     *
+     * This function should only be called once per workflow execution.
+     *
+     * @param stepResult The result of the workflow execution.
+     */
     const setResult = (stepResult: Omit<ExecuteOutput, 'metadata'>): void => {
-      resolveEarlyExit();
+      concludeExecution();
```
Comment (rifont) [inline packages/framework/src/client.ts:376]: Renaming the early exit `Promise` and `resolve` method semantically

File: packages/framework/src/client.ts [hunk 3]
```diff
@@ -401,7 +421,7 @@ export class Client {
 
     const elapsedSeconds = endTime[0];
     const elapsedNanoseconds = endTime[1];
-    const elapsedTimeInMilliseconds = elapsedSeconds * 1000 + elapsedNanoseconds / 1_000_000;
+    const elapsedTimeInMilliseconds = elapsedSeconds * 1_000 + elapsedNanoseconds / 1_000_000;
```
Comment (rifont) [inline packages/framework/src/client.ts:424]: Consistent use of `_` delimiters for numbers.

File: packages/framework/src/client.ts [hunk 4]
```diff
@@ -603,8 +623,8 @@ export class Client {
           });
 
           return {
-            outputs: validatedResult,
-            providers: await this.executeProviders(event, step, validatedResult),
+            outputs: validatedOutput,
+            providers: await this.executeProviders(event, step, validatedOutput),
```
Comment (rifont) [inline packages/framework/src/client.ts:627]: This was a typo earlier.

File: packages/framework/src/client.ts [hunk 5]
```diff
@@ -347,13 +355,25 @@ export class Client {
       providers: {},
       options: { skip: false },
     };
-    let resolveEarlyExit: (value?: unknown) => void;
-    const earlyExitPromise = new Promise((resolve) => {
-      resolveEarlyExit = resolve;
-    });
 
+    let concludeExecution: (value?: unknown) => void;
```
Comment (antonjoel82) [inline packages/framework/src/client.ts:359]: 👏 **praise:** Nice name change and thanks for the comment!‏

File: packages/framework/src/client.ts (hunk 3 above)
Comment (antonjoel82) [reply packages/framework/src/client.ts:424]: ❓ **question:** Is this something we plan to do across the code base?‏

**Review by antonjoel82 — APPROVED** [review]
🎉

File: packages/framework/src/client.ts (hunk 3 above)
Comment (rifont) [reply packages/framework/src/client.ts:424]: That would be my preference. I'm going to add an item to our Eng-Coffee-Chats for us to decide on it.

**Pull request merged**